src/icp_canister_backend/src/lib.rs

```diff
@@ -128,4 +134,53 @@ pub async fn deposit(user: Principal, timelock: u64) -> Result<(), types::Deposi
     Ok(())
 }
 
+#[ic_cdk::update]
+pub async fn withdraw(deposit_id: u64) -> Result<Nat, types::WithdrawError> {
+    let caller = ic_cdk::api::caller();
+    let now = ic_cdk::api::time();
+
+    let deposit = DEPOSITS.with(|deposits| deposits.borrow_mut().remove(&deposit_id));
+    let deposit = match deposit {
+        Some(d) => d,
+        None => return Err(types::WithdrawError::NoDeposit),
+    };
+
+    if deposit.principal != caller {
+        DEPOSITS.with(|deposits| deposits.borrow_mut().insert(deposit_id, deposit));
```

Comment:
These inserts make unnecessary code duplication.
Instead
First, get the immutable deposit struct without removal
Then run validation on it
Then remove it when it is known that it should be removed.
..other logic continues..